File: backend/internal/adapter/server/project.go
package server

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/x-sushant-x/connective/internal/core/domain"
	"github.com/x-sushant-x/connective/internal/core/dto"
	"github.com/x-sushant-x/connective/internal/core/port"
)

type ProjectHandler struct {
	svc port.ProjectService
}

func NewProjectHandler(svc port.ProjectService) *ProjectHandler {
	return &ProjectHandler{
		svc,
	}
}

// HandleCreateProject godoc
// @Security BearerAuth
// @Summary Create a new project.
// @Description Create a new project.
// @Tags Project
// @Accept  json
// @Produce  json
// @Param  request body dto.CreateProjectRequest true "Create Project Body"
// @Success 200 {object} APIResponse{data=domain.Project}
// @Router /api/v1/project/create [post]
func (ph *ProjectHandler) HandleCreateProject(ctx *gin.Context) {
	var req *dto.CreateProjectRequest

	if err := ctx.BindJSON(&req); err != nil {
		BadRequest(ctx, "Bad Request")
		return
	}

	currentUser, ok := ctx.Get("currentUser")
	if !ok {
		Unauthorized(ctx)
		return
	}

	user := currentUser.(*domain.User)

	newProject, err := ph.svc.CreateProject(ctx, user, req)
	if err != nil {
		BadRequest(ctx, err.Error())
		return
	}

	Success(ctx, newProject, "Project Created", http.StatusOK)
}

// HandleGetAllProjects godoc
// @Summary Get all projects of a user.
// @Description Get all projects of a user.
// @Tags Project
// @Accept  json
// @Produce  json
// @Success 200 {object} APIResponse{data=[]domain.Project}
// @Router /api/v1/project/get-all [get]
func (ph *ProjectHandler) HandleGetAllProjects(ctx *gin.Context) {
	currentUser, ok := ctx.Get("currentUser")
	if !ok {
		Unauthorized(ctx)
		return
	}

	user := currentUser.(*domain.User)

	resp, err := ph.svc.GetByOwner(ctx, user.ID)
	if err != nil {
		BadRequest(ctx, err.Error())
		return
	}

	Success(ctx, resp, "Projects List Fetched", http.StatusOK)
}

File: dashboard/src/api/project.ts
import { Endpoints } from "@/constants";
import axios, { isAxiosError } from "axios";
import type { CreateProjectRequest, CreateProjectResponse, Project } from "./dto/project";
import type { APIResponse } from "./dto/api_response";


export const CreateProject = async (name: string) => {
    const req: CreateProjectRequest = { name };

    try {
        const resp = await axios.post<APIResponse<CreateProjectResponse>>(
            Endpoints.Project.Create,
            req,
            {
                withCredentials: true,
            },
        );
        if (resp.data.error) {
            throw new Error(resp.data.error);
        }

        return resp.data;
    } catch (error) {
        if (isAxiosError(error)) {
            const backendMessage =
                error.response?.data?.error ||
                error.response?.data?.message ||
                error.message;

            throw new Error(backendMessage);
        }

        throw new Error("Unexpected error");
    }
};

export const getAllProjects = async (): Promise<Project[]> => {
    try {
        const resp = await axios.get<APIResponse<Project[]>>(Endpoints.Project.GetAllProjects, {
            withCredentials: true,
        });

        if (resp.data.error) {
            throw new Error(resp.data.error);
        }

        return resp.data.data;
    } catch (err) {
        if (isAxiosError(err)) {
            const msg = err.response?.data?.error || err.message;
            throw new Error(msg);
        }

        throw new Error("Unexpected Error");
    }
};